main/src/library/Vector.flix

```diff
@@ -0,0 +1,1389 @@
+        if (length(v) < 1)
+            empty()
+        else
+            if (n < 0)
+                rotateLeftHelper(Int32.abs(n), v)
+            else
+                rotateRightHelper(n, v)
+
+    ///
+    /// Helper function for `rotateRight` and `rotateLeft`.
+    ///
+    /// Precondition: `n` must be positive.
+    ///
+    /// This is an explicit helper to avoid code duplication.
+    ///
+    def rotateRightHelper(n: Int32, v: Vector[a]): Vector[a] =
+        let len = length(v);
+        let f = ix -> {let readIx = Int32.mod(ix - n, len); get(readIx, v)};
+        init(f, len)
+
+    ///
+    /// Returns a copy of `v` with the element at index `i` replaced by `x`.
+    ///
+    /// Returns a copy of `v` if `i < 0` or `i > length(xs)-1`.
+    ///
+    pub def update(i: Int32, x: a, v: Vector[a]): Vector[a] =
+        let f = ix -> if (ix == i) x else get(ix, v);
+        init(f, length(v))
+
+    ///
+    /// Returns `b` with the `n` elements starting at index `i` replaced with the elements of `a`.
+    ///
+    /// If any of the indices `i, i+1, i+2, ... , i+n-1` are out of range in `b` then no patching is done at these indices.
+    /// If `a` becomes depleted then no further patching is done.
+    /// If patching occurs at index `i+j` in `b`, then the element at index `j` in `a` is used.
+    ///
+    pub def patch(i: Int32, n: Int32, a: Vector[a], b: Vector[a]): Vector[a] =
+        let len1 = length(a);
+        let size = if (n > len1) len1 else n;
+        let sub = slice(start = 0, end = size, a);
+        updateSequence(i, sub, b)
+
+    ///
+    /// Returns a copy of `v` with `sep` inserted between every two adjacent elements.
+    ///
+    pub def intersperse(sep: a, v: Vector[a]): Vector[a] =
+        let len1 = length(v);
+        let len2 = len1 + len1 - 1;
+        let f = ix -> match ix {
+            case 0                         => get(0, v)
+            case n if Int32.rem(n, 2) != 0 => sep
+            case n                         => {let i = n / 2; get(i, v)}
+        };
+        init(f, len2)
+
+    ///
+    /// Returns the concatenation of the elements in `vs` with the elements
+    /// of `sep` inserted between every two adjacent elements.
+    ///
+    pub def intercalate(sep: Vector[a], vs: Vector[Vector[a]]): Vector[a] = region rc {
+        let count = length(vs);
+        let sepLength = length(sep);
+        let sepCount = if (count < 2) 0 else count - 1;
+        let len = sumLengths(vs) + (sepCount * sepLength);
+        let pos = ref 0 @ rc;
+        let sepArr = toArray(rc, sep);
+        buildVector(rc, len, arr -> {
+            forEachWithIndex((i, v) ->
+                if (i == 0) {
+                    let a1 = toArray(rc, v);
+                    Array.updateSequence!(0, a1, arr);
+                    pos := length(v)
+                } else {
+                    let ix = deref pos;
+                    Array.updateSequence!(ix, sepArr, arr);
```

Comment:
Same comment about not using Array/updateSequence.

Reply:
There's now a helper function in Vector:

> def arrayUpdateSeqV!(i: Int32, sub: Vector[a], arr: Array[a, r]): Unit \ Write(r)  =

This is `Array.updateSequence` except `sub` is used directly as a vector and not converted into an array.